File: backend/app/services/alert_service.py
import uuid
from typing import List, Optional
from datetime import datetime, timezone
from sqlalchemy.orm import Session
from app.models.alert import Alert
from app.models.field import Field
from app.schemas.alert import AlertCreate, AlertResponse


class AlertService:
    @staticmethod
    def get_by_farm(db: Session, farm_id: str, status: Optional[str] = None) -> List[AlertResponse]:
        query = db.query(Alert).filter(Alert.farm_id == farm_id)
        if status:
            query = query.filter(Alert.status == status.upper())
        alerts = query.order_by(Alert.created_at.desc()).all()

        results = []
        for a in alerts:
            field = db.query(Field).filter(Field.id == a.field_id).first() if a.field_id else None
            resp = AlertResponse.model_validate(a)
            resp.field_name = field.name if field else "All Fields"
            results.append(resp)
        return results

    @staticmethod
    def resolve_alert(db: Session, alert_id: str) -> Optional[AlertResponse]:
        alert = db.query(Alert).filter(Alert.id == alert_id).first()
        if not alert:
            return None
        alert.status = "RESOLVED"
        alert.resolved_at = datetime.now(timezone.utc)
        db.commit()
        db.refresh(alert)
        field = db.query(Field).filter(Field.id == alert.field_id).first() if alert.field_id else None
        resp = AlertResponse.model_validate(alert)
        resp.field_name = field.name if field else "All Fields"
        return resp

    @staticmethod
    def create_alert(db: Session, alert_in: AlertCreate) -> AlertResponse:
        now = datetime.now(timezone.utc)
        alert = Alert(
            id=alert_in.id or f"alert_{uuid.uuid4().hex[:8]}",
            farm_id=alert_in.farm_id,
            field_id=alert_in.field_id,
            sensor_type=alert_in.sensor_type,
            severity=alert_in.severity,
            title=alert_in.title,
            message=alert_in.message,
            recommendation=alert_in.recommendation,
            status="ACTIVE",
            created_at=now,
        )
        db.add(alert)
        db.commit()
        db.refresh(alert)
        return AlertResponse.model_validate(alert)

File: frontend/src/services/alertService.js
/**
 * @file alertService.js
 * Alert & Action Center Service.
 * Manages actionable farmer alerts, resolution states, and snoozes.
 */

import { INITIAL_ALERTS } from '../data/seedAlerts';
import { api } from './apiClient';

const ALERTS_STORAGE_KEY = 'acrovision_alerts_data';

function loadLocalAlerts() {
  try {
    const data = localStorage.getItem(ALERTS_STORAGE_KEY);
    if (data) return JSON.parse(data);
  } catch {
    // Fallback
  }
  return INITIAL_ALERTS;
}

function saveLocalAlerts(alerts) {
  try {
    localStorage.setItem(ALERTS_STORAGE_KEY, JSON.stringify(alerts));
  } catch {
    // Local storage error
  }
}

function normalizeAlert(a) {
  return {
    id: a.id,
    farmId: a.farm_id || a.farmId || 'farm_01',
    fieldId: a.field_id || a.fieldId,
    fieldName: a.field_name || a.fieldName || 'Monitored Field',
    title: a.title,
    description: a.message || a.description,
    recommendedAction: a.recommendation || a.recommendedAction,
    whenToDo: a.when_to_do || a.whenToDo || 'Today',
    severity: a.severity || 'WATCH',
    category: a.sensor_type || a.category || 'irrigation',
    isRead: false,
    isResolved: a.status === 'RESOLVED' || a.isResolved || false,
    createdAt: a.created_at || new Date().toISOString(),
    resolvedAt: a.resolved_at || a.resolvedAt || null,
    actions: [
      { id: `act-${a.id}`, label: 'Mark as Done', variant: 'primary', type: 'RESOLVE' },
    ],
  };
}

export const alertService = {
  /**
   * Get all alerts (active or resolved) from backend API or local fallback.
   */
  async getAlerts(farmId = 'farm_01') {
    try {
      const backendAlerts = await api.get(`/farms/${farmId}/alerts`);
      if (Array.isArray(backendAlerts)) {
        const normalized = backendAlerts.map(normalizeAlert);
        saveLocalAlerts(normalized);
        return normalized;
      }
    } catch {
      // Backend offline fallback
    }
    return loadLocalAlerts();
  },

  /**
   * Get active alerts count.
   */
  async getActiveAlertsCount(farmId = 'farm_01') {
    const alerts = await this.getAlerts(farmId);
    return alerts.filter((a) => !a.isResolved).length;
  },

  /**
   * Mark alert as resolved / done.
   */
  async resolveAlert(alertId) {
    try {
      const res = await api.patch(`/alerts/${alertId}/resolve`, {});
      return normalizeAlert(res);
    } catch {
      // Local fallback
    }
    const alerts = loadLocalAlerts();
    const index = alerts.findIndex((a) => a.id === alertId);
    if (index !== -1) {
      alerts[index].isResolved = true;
      alerts[index].isRead = true;
      alerts[index].resolvedAt = new Date().toISOString();
      saveLocalAlerts(alerts);
      return alerts[index];
    }
    return null;
  },

  /**
   * Mark alert as read.
   */
  async markAsRead(alertId) {
    const alerts = loadLocalAlerts();
    const index = alerts.findIndex((a) => a.id === alertId);
    if (index !== -1) {
      alerts[index].isRead = true;
      saveLocalAlerts(alerts);
      return alerts[index];
    }
    return null;
  },

  /**
   * Snooze alert for a given period.
   */
  async snoozeAlert(alertId, hours = 2) {
    const alerts = loadLocalAlerts();
    const index = alerts.findIndex((a) => a.id === alertId);
    if (index !== -1) {
      alerts[index].isRead = true;
      alerts[index].whenToDo = `Snoozed for ${hours}h`;
      saveLocalAlerts(alerts);
      return alerts[index];
    }
    return null;
  },

  /**
   * Create an alert dynamically.
   */
  async createAlert(alertData) {
    const payload = {
      farm_id: alertData.farmId || 'farm_01',
      field_id: alertData.fieldId,
      sensor_type: alertData.category || 'soil_moisture',
      severity: alertData.severity || 'ACTION_NEEDED',
      title: alertData.title,
      message: alertData.description || alertData.message,
      recommendation: alertData.recommendedAction || alertData.recommendation,
    };

    try {
      const created = await api.post('/alerts', payload);
      const normalized = normalizeAlert(created);
      const local = loadLocalAlerts();
      saveLocalAlerts([normalized, ...local]);
      return normalized;
    } catch {
      const alerts = loadLocalAlerts();
      const newAlert = {
        id: `alert-${Date.now()}`,
        farmId: alertData.farmId || 'farm_01',
        fieldId: alertData.fieldId,
        fieldName: alertData.fieldName,
        title: alertData.title,
        description: alertData.description,
        recommendedAction: alertData.recommendedAction,
        whenToDo: alertData.whenToDo || 'Today',
        severity: alertData.severity || 'WATCH',
        category: alertData.category || 'irrigation',
        isRead: false,
        isResolved: false,
        createdAt: new Date().toISOString(),
        resolvedAt: null,
        actions: [
          { id: `act-${Date.now()}`, label: 'Mark as Done', variant: 'primary', type: 'RESOLVE' },
        ],
      };
      saveLocalAlerts([newAlert, ...alerts]);
      return newAlert;
    }
  },
};

